src/main/scala/mesosphere/marathon/api/v2/serialization/ContainerSerializer.scala

```diff
@@ -0,0 +1,165 @@
+    val maybeDocker = if (proto.hasDocker) Some(Docker(proto.getDocker)) else None
+    Container(
+      `type` = proto.getType,
+      volumes = proto.getVolumesList.asScala.map(Container.Volume(_)).to[Seq],
+      docker = maybeDocker
+    )
+  }
+
+  def toMesos(container: Container): mesos.Protos.ContainerInfo = {
+    // we can only serialize DockerVolumes into a Mesos Protobuf.
+    // PersistentVolumes and ExternalVolumes are handled differently
+    val serializedVolumes = container.volumes.collect { case dv: DockerVolume => VolumeSerializer.toMesos(dv) }
+    val builder = mesos.Protos.ContainerInfo.newBuilder
+      .setType(container.`type`)
+      .addAllVolumes(serializedVolumes.asJava)
+    container.docker.foreach { d => builder.setDocker(DockerSerializer.toMesos(d)) }
+    builder.build
+  }
+
+  /**
+    * Lossy conversion for backwards compatibility with deprecated
+    * container representation.
+    */
+  def fromMesos(proto: mesos.Protos.CommandInfo.ContainerInfo): Container =
+    Container(
+      `type` = mesos.Protos.ContainerInfo.Type.DOCKER,
+      docker = Some(Docker(proto.getImage))
+    )
+
+  /**
+    * Lossy conversion for backwards compatibility with deprecated
+    * container representation.
+    */
+  def fromMesos(proto: Protos.ContainerInfo): Container =
+    Container(
+      `type` = mesos.Protos.ContainerInfo.Type.DOCKER,
+      docker = Some(Docker(proto.getImage.toStringUtf8))
+    )
+
+}
+
+object VolumeSerializer {
+
+  def toProto(volume: Volume): Protos.Volume = volume match {
+    case p: PersistentVolume =>
+      Protos.Volume.newBuilder()
+        .setContainerPath(p.containerPath)
+        .setPersistent(PersistentVolumeInfoSerializer.toProto(p.persistent))
+        .setMode(p.mode)
+        .build()
+
+    case d: DockerVolume =>
+      Protos.Volume.newBuilder()
+        .setContainerPath(d.containerPath)
+        .setHostPath(d.hostPath)
+        .setMode(d.mode)
+        .build()
+  }
+
+  /** Only DockerVolumes can be serialized into a Mesos Protobuf */
+  def toMesos(volume: DockerVolume): mesos.Protos.Volume =
+    mesos.Protos.Volume.newBuilder
+      .setContainerPath(volume.containerPath)
+      .setHostPath(volume.hostPath)
+      .setMode(volume.mode)
+      .build
+}
+
+object PersistentVolumeInfoSerializer {
+  def toProto(info: PersistentVolumeInfo): Protos.PersistentVolumeInfo =
+    Protos.PersistentVolumeInfo.newBuilder()
+      .setSize(info.size)
+      .build()
+}
+
+object DockerSerializer {
+  def toProto(docker: Container.Docker): Protos.ExtendedContainerInfo.DockerInfo = {
+    val builder = Protos.ExtendedContainerInfo.DockerInfo.newBuilder
+
+    builder.setImage(docker.image)
+
+    docker.network foreach builder.setNetwork
+
+    docker.portMappings.foreach { pms =>
+      builder.addAllPortMappings(pms.map(PortMappingSerializer.toProto).asJava)
+    }
+
+    builder.setPrivileged(docker.privileged)
```

Comment:
Did that where it's cheap – setNetwork doesn't allow null so I can't chain that call, and for portMappings it'd be a little ugly because I'd have to fold over the Option and pass an empty list if there aren't any portMappings – easier to read is it is imho.